src/persist-client/src/impl/encoding.rs

```diff
@@ -87,140 +92,169 @@ where
+    // so that a change to the crate can't break us. The compiler statically
+    // checks that they match because this const is used in the return type.
+    // (MD5 is not going to change, so this is all a moot point, but still
+    // better to be defensive.)
+    const CHECKSUM_LEN: usize = 16;
+
+    fn decode_checksummed(buf: &[u8]) -> Result<ProtoStateRollup, String> {
+        if buf.len() < Self::CHECKSUM_LEN {
+            return Err("missing/incomplete checksum".into());
+        }
+        let checksum_start = buf.len() - Self::CHECKSUM_LEN;
+        let (buf, checksum) = (&buf[..checksum_start], &buf[checksum_start..]);
+        if checksum != &Self::md5_checksum(&buf) {
+            return Err("checksum mismatch".into());
+        }
+
+        ProtoStateRollup::decode(buf).map_err(|err| err.to_string())
+    }
+
+    fn md5_checksum(buf: &[u8]) -> [u8; Self::CHECKSUM_LEN] {
+        let mut h = Md5::new();
+        h.update(&buf);
+        h.finalize().into()
     }
 }
 
-impl<K, V, T, D> From<&State<K, V, T, D>> for StateRollupMeta
+impl<K, V, T, D> RustType<ProtoStateRollup> for State<K, V, T, D>
 where
     K: Codec,
     V: Codec,
     T: Timestamp + Lattice + Codec64,
     D: Codec64,
 {
-    fn from(x: &State<K, V, T, D>) -> Self {
-        StateRollupMeta {
-            shard_id: x.shard_id,
-            seqno: x.seqno,
+    fn into_proto(&self) -> ProtoStateRollup {
+        ProtoStateRollup {
+            shard_id: self.shard_id.into_proto(),
+            seqno: self.seqno.into_proto(),
             key_codec: K::codec_name(),
             val_codec: V::codec_name(),
             ts_codec: T::codec_name(),
             diff_codec: D::codec_name(),
-            readers: x
+            readers: self
                 .collections
                 .readers
                 .iter()
-                .map(|(id, cap)| (id.clone(), (&cap.since).into(), cap.seqno))
+                .map(|(id, cap)| ProtoReader {
+                    reader_id: id.into_proto(),
+                    since: Some(cap.since.into_proto()),
+                    seqno: cap.seqno.into_proto(),
+                })
                 .collect(),
-            trace: (&x.collections.trace).into(),
+            trace: Some(self.collections.trace.into_proto()),
         }
     }
+
+    fn from_proto(proto: ProtoStateRollup) -> Result<Self, TryFromProtoError> {
+        State::try_from(proto).map(|x| {
+            // Turn this into a TryFromProtoError?
+            x.expect("codec mismatch")
+        })
+    }
 }
 
-impl<K, V, T, D> TryFrom<&StateRollupMeta> for State<K, V, T, D>
+impl<K, V, T, D> State<K, V, T, D>
 where
     K: Codec,
     V: Codec,
     T: Timestamp + Lattice + Codec64,
     D: Codec64,
 {
-    type Error = InvalidUsage<T>;
-
-    fn try_from(x: &StateRollupMeta) -> Result<Self, Self::Error> {
+    fn try_from(x: ProtoStateRollup) -> Result<Result<Self, InvalidUsage<T>>, TryFromProtoError> {
```

Comment:
Side note: I don't love that we're doing the codec checking here. This might be nicer when we remove the types from `Machine`/`State` and do the checking in the handles.